data/transactions/logic/debugger.go

```diff
@@ -34,14 +34,157 @@ import (
 )
 
 // DebuggerHook functions are called by eval function during TEAL program execution
-// if provided
+// if provided. The interface is empty because none of the hooks are required by default.
+//
+// See `debuggerBeforeTxnHook`, `debuggerBeforeAppEvalHook`, etc. for supported
+// interface methods and refer to the lifecycle graph within the DebuggerHook interface definition for
+// the sequence in which hooks are called.
+//
+// NOTE: Debugger hooks are passed by reference to DebugState and EvalParams and are not copies.
+// It is therefore the responsibility of the debugger hooks to not modify the state of the structs
+// passed to them. Additionally, hooks are responsible for copying the information
+// they need from the state and params structs. No guarantees are made that the referenced state
+// will not change between hook calls. This decision was made in an effort to reduce the performance
+// impact of the debugger hooks.
 type DebuggerHook interface {
-	// Register is fired on program creation
-	Register(state *DebugState) error
-	// Update is fired on every step
-	Update(state *DebugState) error
-	// Complete is called when the program exits
-	Complete(state *DebugState) error
+
+	// LIFECYCLE GRAPH
+	// ┌────────────────────────────────────────────────┐
+	// │ Transaction Evaluation                         │
+	// ├────────────────────────────────────────────────┤
+	// │ > BeforeTxn                                    │
```

Comment:
Have you thought about if/how the interface can support LogicSig evaluation as well?

Reply:
I had not. Glad someone was thinking about that 😅 . I've since added implementation details.